File: Sources/IrohaCore/LearningDictionary.swift
import Foundation

/// ユーザが変換を修正したときに記録する1エントリ。
public struct LearningEntry: Codable, Sendable, Hashable {

    public enum Kind: String, Codable, Sendable {
        case sentence   // 入力全体（読み全体 → 確定文字列）
        case segment    // 文節（読み → 変換結果、直前の文脈つき）
    }

    public var kind: Kind
    /// ひらがなの読み
    public var reading: String
    /// 確定された変換結果
    public var result: String
    /// この入力内で直前までに確定していた文字列の末尾（最大`LearningDictionary.contextLength`文字）。
    /// 入力の先頭の文節なら空。`kind == .sentence`では使わない
    public var leftContext: String
    public var updatedAt: Date

    public init(
        kind: Kind, reading: String, result: String, leftContext: String = "",
        updatedAt: Date = Date()
    ) {
        self.kind = kind
        self.reading = reading
        self.result = result
        self.leftContext = leftContext
        self.updatedAt = updatedAt
    }
}

/// 変換時に参照する学習結果の不変スナップショット。
///
/// 同じ読みでも文中の位置によって正解が違う（「きしゃのきしゃ」→「記者の貴社」）ため、
/// 文節の学習は「直前までに確定した文字列」を条件に付けて記録・適用する。
public struct LearningDictionary: Sendable {

    /// 文脈として覚える文字数
    public static let contextLength = 6
    /// 文中の一部として当てはめる最短の読みの長さ（1文字だと無差別に一致して変換が壊れる）
    public static let minimumMatchLength = 2

    public static let empty = LearningDictionary(entries: [])

    public let entries: [LearningEntry]
    /// 読み全体 → 確定文字列
    private let sentences: [String: LearningEntry]
    /// 読み → 文節エントリ（新しい順）
    private let segments: [String: [LearningEntry]]
    private let maxSegmentReadingLength: Int

    public init(entries: [LearningEntry]) {
        self.entries = entries
        var sentences: [String: LearningEntry] = [:]
        var segments: [String: [LearningEntry]] = [:]
        var maxLength = 0
        for entry in entries.sorted(by: { $0.updatedAt > $1.updatedAt }) {
            guard !entry.reading.isEmpty, !entry.result.isEmpty else { continue }
            switch entry.kind {
            case .sentence:
                // 同じ読みなら新しい方を採る（並べ替え済みなので先勝ち）
                if sentences[entry.reading] == nil { sentences[entry.reading] = entry }
            case .segment:
                segments[entry.reading, default: []].append(entry)
                maxLength = max(maxLength, entry.reading.count)
            }
        }
        self.sentences = sentences
        self.segments = segments
        self.maxSegmentReadingLength = maxLength
    }

    public var isEmpty: Bool { sentences.isEmpty && segments.isEmpty }

    // MARK: - 参照

    /// 入力全体が過去に確定した読みと完全一致するならその確定文字列
    public func sentence(forReading reading: String) -> String? {
        sentences[reading]?.result
    }

    /// 位置`index`から始まる学習済み文節が適用され得るか（文脈条件込みの安価な事前判定）。
    ///
    /// 直前の未変換部分を先にエンジンへ渡して文脈を確定させる必要があるので、
    /// 「変換してみるまで分からない」条件（空でない`leftContext`）はここでは真として扱う
    public func mayMatch(_ characters: [Character], from index: Int, atStart: Bool) -> Bool {
        forEachCandidateReading(characters, from: index) { reading in
            guard let candidates = segments[reading] else { return false }
            // 文脈なしで覚えたものは入力の先頭でしか使わない
            return candidates.contains { !$0.leftContext.isEmpty || atStart }
        }
    }

    /// 位置`index`から始まる学習済み文節のうち、文脈条件を満たす最長のもの
    public func bestMatch(
        _ characters: [Character], from index: Int, leftContext: String, atStart: Bool
    ) -> LearningEntry? {
        var found: LearningEntry?
        _ = forEachCandidateReading(characters, from: index) { reading in
            guard let candidates = segments[reading] else { return false }
            // 新しい順に見て、最初に文脈が合ったものを採る
            guard let match = candidates.first(where: {
                applies($0, leftContext: leftContext, atStart: atStart)
            }) else { return false }
            found = match
            return true
        }
        return found
    }

    /// 文脈条件の判定
    public func applies(_ entry: LearningEntry, leftContext: String, atStart: Bool) -> Bool {
        entry.leftContext.isEmpty ? atStart : leftContext.hasSuffix(entry.leftContext)
    }

    /// 候補ウィンドウ用: この読みで学習済みの変換結果を、文脈が合うものから新しい順に返す
    public func learnedResults(forReading reading: String, leftContext: String) -> [String] {
        var applicable: [String] = []
        var others: [String] = []
        for entry in segments[reading] ?? [] {
            // 候補ウィンドウは明示的な操作なので、文脈が合わないものも後ろに残す
            if applies(entry, leftContext: leftContext, atStart: leftContext.isEmpty) {
                applicable.append(entry.result)
            } else {
                others.append(entry.result)
            }
        }
        if let sentence = sentences[reading]?.result {
            applicable.insert(sentence, at: 0)
        }
        var results: [String] = []
        for result in applicable + others where !results.contains(result) {
            results.append(result)
        }
        return results
    }

    // MARK: - 内部

    /// 位置`index`から始まる読みを長い方から順に`body`へ渡す。`body`が真を返したら打ち切る
    private func forEachCandidateReading(
        _ characters: [Character], from index: Int, _ body: (String) -> Bool
    ) -> Bool {
        var length = min(maxSegmentReadingLength, characters.count - index)
        while length >= Self.minimumMatchLength {
            if body(String(characters[index..<(index + length)])) { return true }
            length -= 1
        }
        return false
    }
}

File: windows/iroha-core/src/learning_dictionary.cpp
#include "iroha/learning_dictionary.h"

#include <algorithm>

namespace iroha {

namespace {

bool HasSuffix(const std::u32string& text, const std::u32string& suffix) {
    return text.size() >= suffix.size() &&
           text.compare(text.size() - suffix.size(), suffix.size(), suffix) == 0;
}

} // namespace

LearningDictionary::LearningDictionary(std::vector<LearningEntry> entries)
    : entries_(std::move(entries)) {
    std::vector<const LearningEntry*> sorted;
    sorted.reserve(entries_.size());
    for (const LearningEntry& entry : entries_) sorted.push_back(&entry);
    // 新しい順（同時刻は元の並びを保つ）
    std::stable_sort(sorted.begin(), sorted.end(),
                     [](const LearningEntry* a, const LearningEntry* b) {
                         return a->updatedAt > b->updatedAt;
                     });
    for (const LearningEntry* entry : sorted) {
        if (entry->reading.empty() || entry->result.empty()) continue;
        switch (entry->kind) {
            case LearningEntry::Kind::Sentence:
                // 同じ読みなら新しい方を採る（並べ替え済みなので先勝ち）
                sentences_.emplace(entry->reading, *entry);
                break;
            case LearningEntry::Kind::Segment:
                segments_[entry->reading].push_back(*entry);
                maxSegmentReadingLength_ =
                    std::max(maxSegmentReadingLength_, entry->reading.size());
                break;
        }
    }
}

const std::u32string* LearningDictionary::Sentence(const std::u32string& reading) const {
    auto it = sentences_.find(reading);
    return it != sentences_.end() ? &it->second.result : nullptr;
}

bool LearningDictionary::MayMatch(const std::u32string& characters, size_t index,
                                  bool atStart) const {
    // 位置indexから始まる読みを長い方から順に調べる
    size_t length = std::min(maxSegmentReadingLength_, characters.size() - index);
    while (length >= static_cast<size_t>(kMinimumMatchLength)) {
        auto it = segments_.find(characters.substr(index, length));
        if (it != segments_.end()) {
            // 文脈なしで覚えたものは入力の先頭でしか使わない
            const bool usable = std::any_of(
                it->second.begin(), it->second.end(),
                [atStart](const LearningEntry& e) {
                    return !e.leftContext.empty() || atStart;
                });
            if (usable) return true;
        }
        --length;
    }
    return false;
}

const LearningEntry* LearningDictionary::BestMatch(const std::u32string& characters,
                                                   size_t index,
                                                   const std::u32string& leftContext,
                                                   bool atStart) const {
    size_t length = std::min(maxSegmentReadingLength_, characters.size() - index);
    while (length >= static_cast<size_t>(kMinimumMatchLength)) {
        auto it = segments_.find(characters.substr(index, length));
        if (it != segments_.end()) {
            // 新しい順に見て、最初に文脈が合ったものを採る
            for (const LearningEntry& entry : it->second) {
                if (Applies(entry, leftContext, atStart)) return &entry;
            }
        }
        --length;
    }
    return nullptr;
}

bool LearningDictionary::Applies(const LearningEntry& entry,
                                 const std::u32string& leftContext,
                                 bool atStart) const {
    return entry.leftContext.empty() ? atStart
                                     : HasSuffix(leftContext, entry.leftContext);
}

std::vector<std::u32string> LearningDictionary::LearnedResults(
    const std::u32string& reading, const std::u32string& leftContext) const {
    std::vector<std::u32string> applicable;
    std::vector<std::u32string> others;
    auto it = segments_.find(reading);
    if (it != segments_.end()) {
        for (const LearningEntry& entry : it->second) {
            // 候補ウィンドウは明示的な操作なので、文脈が合わないものも後ろに残す
            if (Applies(entry, leftContext, leftContext.empty())) {
                applicable.push_back(entry.result);
            } else {
                others.push_back(entry.result);
            }
        }
    }
    auto sentence = sentences_.find(reading);
    if (sentence != sentences_.end()) {
        applicable.insert(applicable.begin(), sentence->second.result);
    }
    std::vector<std::u32string> results;
    for (const auto& list : {applicable, others}) {
        for (const auto& result : list) {
            if (std::find(results.begin(), results.end(), result) == results.end()) {
                results.push_back(result);
            }
        }
    }
    return results;
}

} // namespace iroha
